AQ0X_TESTER_CONFIG/doxygen/man/man3/emu_data_editor.3
.TH "emu_data_editor" 3 "Thu Oct 30 2014" "Version V0.0" "AQ0X" \" -*- nroff -*-
.ad l
.nh
.SH NAME
emu_data_editor \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <emu_data_editor\&.h>\fP
.PP
Inherits \fBQMainWindow\fP\&.
.SS "Public Slots"

.in +1c
.ti -1c
.RI "void \fBupdate_curve\fP ()"
.br
.ti -1c
.RI "void \fBselection\fP (bool clear)"
.br
.in -1c
.SS "Signals"

.in +1c
.ti -1c
.RI "void \fBedit_point_selection\fP (bool status)"
.br
.ti -1c
.RI "void \fBedit_point_deselection\fP (bool status)"
.br
.ti -1c
.RI "void \fBadjust_value\fP (\fBInt16\fP val)"
.br
.ti -1c
.RI "void \fBset_value\fP (\fBInt16\fP val, \fBUint16\fP waveform)"
.br
.ti -1c
.RI "void \fBaction_set_value\fP (double val, \fBUint8\fP type)"
.br
.ti -1c
.RI "void \fBupdate_curve_signal\fP ()"
.br
.ti -1c
.RI "void \fBupdate_value\fP ()"
.br
.ti -1c
.RI "void \fBset_status\fP (\fBUint8\fP select)"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBemu_data_editor\fP (QString type, QWidget *parent=0)"
.br
.ti -1c
.RI "\fB~emu_data_editor\fP ()"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "QString \fBwin_title\fP"
.br
.ti -1c
.RI "\fBUint8\fP \fBsignal_type\fP"
.br
.ti -1c
.RI "double \fBsignal_freq\fP"
.br
.ti -1c
.RI "double \fBsignal_phase\fP"
.br
.ti -1c
.RI "double \fBsignal_amplitude\fP"
.br
.ti -1c
.RI "QVector< \fBemu_data_edit_point_t\fP > \fBedit_point_sets\fP"
.br
.ti -1c
.RI "QFourierTransformer \fBtransformer\fP"
.br
.ti -1c
.RI "QVector< double > \fBemu_data_yvalue\fP"
.br
.ti -1c
.RI "QVector< double > \fBemu_data_xvalue\fP"
.br
.ti -1c
.RI "float \fBfft_result\fP [\fBFFT_SIZE\fP]"
.br
.ti -1c
.RI "float \fBfft_inputs\fP [\fBFFT_SIZE\fP]"
.br
.ti -1c
.RI "QVector< double > \fBemu_data_freq_value\fP"
.br
.ti -1c
.RI "QVector< double > \fBemu_data_spectrum_magnitude\fP"
.br
.ti -1c
.RI "float \fBspectrum_magnitude_max\fP"
.br
.ti -1c
.RI "float \fBspectrum_magnitude_min\fP"
.br
.ti -1c
.RI "QString \fBdir_default\fP"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "bool \fBeventFilter\fP (QObject *obj, QEvent *ev)"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 19 of file emu_data_editor\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "emu_data_editor::emu_data_editor (QStringtype, QWidget *parent = \fC0\fP)\fC [explicit]\fP"
defalt values 
.PP
Definition at line 12 of file emu_data_editor\&.cpp\&.
.PP
References Ui_emu_data_editor::actionWinEditor, Ui_emu_data_editor::actionWinPlot, Ui_emu_data_editor::centralwidget, CONFIG_DEFAULT_FREQ, dir_default, Ui_emu_data_editor::dockWidgetCurve_plot, Ui_emu_data_editor::dockWidgetCurve_plot_fft, emu_data_freq_value, emu_data_spectrum_magnitude, emu_data_xvalue, emu_data_yvalue, EMU_MODE_DATA_LENGTH, fft_inputs, fft_result, FFT_SIZE, Ui_emu_data_editor::plot_curve, Ui_emu_data_editor::plot_curve_fft, Ui_emu_data_editor::setupUi(), signal_amplitude, signal_freq, signal_phase, signal_type, spectrum_magnitude_max, spectrum_magnitude_min, transformer, WAVEFORM_0_TO_3_3, WAVEFORM_PLUS_MINUS_1_65, and win_title\&.
.PP
.nf
12                                                                :
13     QMainWindow(parent),
14     ui(new Ui::emu_data_editor )
15 {
16 
17     ui->setupUi(this);
18     this->setWindowTitle(type);
19 
20     dir_default = ((MainWindow *)parent)->dir_default;
21 
22     win_title = type;
23     if(type\&.contains("MAGNETIC FIELD")){
24         signal_type = WAVEFORM_PLUS_MINUS_1_65;
25         signal_amplitude = 1\&.65;
26     }else{
27         signal_type = WAVEFORM_0_TO_3_3;
28         signal_amplitude = 3\&.3;
29     }
30 
31     signal_freq      = CONFIG_DEFAULT_FREQ;
32     signal_phase     = 0;
33 
34 
35     ui->centralwidget->setVisible(false);
36     ui->dockWidgetCurve_plot->setWidget(ui->plot_curve);
37     ui->dockWidgetCurve_plot_fft->setWidget(ui->plot_curve_fft);
38 
39     this->addDockWidget(Qt::TopDockWidgetArea,ui->dockWidgetCurve_plot_fft);
40 
41 
42     ui->actionWinEditor->setChecked(true);
43     ui->actionWinPlot->setChecked(true);
44 
45 
46 
47     for (int var = 0; var < EMU_MODE_DATA_LENGTH; ++var) {
48         emu_data_xvalue\&.append(var);
49         emu_data_yvalue\&.append(0);
50     }
51 
52     plot_curve(emu_data_xvalue,emu_data_yvalue);
53 
54     memset(fft_result,0,FFT_SIZE);
55     memset(fft_inputs,0,FFT_SIZE);
56 
57 
58     transformer\&.setSize(FFT_SIZE);
59     if(!transformer\&.setWindowFunction("Rectangular"))
60         qCritical()<<"unable to set window funtion : emu_data_editor";
61 
62     transformer\&.forwardTransform(fft_inputs,fft_result);
63 
64 
65     spectrum_magnitude_max = 0;
66     spectrum_magnitude_min = 0;
67 
68 
69 
70     for (int var = 0; var < FFT_SIZE; ++var) {
71 
72         emu_data_freq_value\&.append(
73                     (var)*signal_freq*
74                     ((signal_type == WAVEFORM_PLUS_MINUS_1_65)?1:2));
75 
76         qDebug()<<"f("<<var<<") = "<< emu_data_freq_value[var]<<";";
77         emu_data_spectrum_magnitude\&.append((double)fft_result[var]);
78 
79         if(fft_result[var] > spectrum_magnitude_max)
80             spectrum_magnitude_max = fft_result[var];
81         if(fft_result[var] < spectrum_magnitude_min)
82             spectrum_magnitude_min = fft_result[var];
83 
84     }
85 
86 
87 
88 
89 
90 
91 
92     plot_curve_fft(emu_data_freq_value,emu_data_spectrum_magnitude);
93     installEventFilter(this);
94     gui_init(EMU_MODE_DATA_LENGTH/2);
95 
96 
97 
98 }
.fi
.SS "emu_data_editor::~emu_data_editor ()"

.PP
Definition at line 101 of file emu_data_editor\&.cpp\&.
.PP
.nf
102 {
103     delete ui;
104 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "void emu_data_editor::action_set_value (doubleval, \fBUint8\fPtype)\fC [signal]\fP"

.PP
Definition at line 277 of file moc_emu_data_editor\&.cpp\&.
.PP
.nf
278 {
279     void *_a[] = { 0, const_cast<void*>(reinterpret_cast<const void*>(&_t1)), const_cast<void*>(reinterpret_cast<const void*>(&_t2)) };
280     QMetaObject::activate(this, &staticMetaObject, 4, _a);
281 }
.fi
.SS "void emu_data_editor::adjust_value (\fBInt16\fPval)\fC [signal]\fP"

.PP
Definition at line 263 of file moc_emu_data_editor\&.cpp\&.
.PP
Referenced by eventFilter()\&.
.PP
.nf
264 {
265     void *_a[] = { 0, const_cast<void*>(reinterpret_cast<const void*>(&_t1)) };
266     QMetaObject::activate(this, &staticMetaObject, 2, _a);
267 }
.fi
.SS "void emu_data_editor::edit_point_deselection (boolstatus)\fC [signal]\fP"

.PP
Definition at line 256 of file moc_emu_data_editor\&.cpp\&.
.PP
Referenced by eventFilter()\&.
.PP
.nf
257 {
258     void *_a[] = { 0, const_cast<void*>(reinterpret_cast<const void*>(&_t1)) };
259     QMetaObject::activate(this, &staticMetaObject, 1, _a);
260 }
.fi
.SS "void emu_data_editor::edit_point_selection (boolstatus)\fC [signal]\fP"

.PP
Definition at line 249 of file moc_emu_data_editor\&.cpp\&.
.PP
Referenced by eventFilter()\&.
.PP
.nf
250 {
251     void *_a[] = { 0, const_cast<void*>(reinterpret_cast<const void*>(&_t1)) };
252     QMetaObject::activate(this, &staticMetaObject, 0, _a);
253 }
.fi
.SS "bool emu_data_editor::eventFilter (QObject *obj, QEvent *ev)\fC [protected]\fP"

.PP
Definition at line 107 of file emu_data_editor\&.cpp\&.
.PP
References adjust_value(), edit_point_deselection(), edit_point_selection(), and update_curve()\&.
.PP
.nf
108 {
109 
110 
111     if(obj!=this)
112         return false;
113 
114         if (event->type() == QEvent::KeyPress) {
115             QKeyEvent *keyEvent = static_cast<QKeyEvent*>(event);
116             switch (keyEvent->key()) {
117             case Qt::Key_Shift:
118                     emit edit_point_selection(true);
119                   return false;
120                  break;
121 
122 
123             case Qt::Key_Plus:
124 
125                     emit adjust_value(10);
126                     update_curve();
127                     return false;
128                  break;
129 
130             case Qt::Key_Minus:
131 
132                     emit adjust_value(-10);
133                     update_curve();
134                     return false;
135                 break;
136             case Qt::Key_Left:
137 
138 
139                     return false;
140                 break;
141 
142             case Qt::Key_Clear:
143                     emit edit_point_deselection(true);
144                     return false;
145                      break;
146 
147 
148             case Qt::Key_Control:
149                     emit edit_point_deselection(true);
150                     return false;
151                  break;
152 
153 
154             case Qt::Key_Right:
155 
156 
157             case Qt::Key_Enter:
158 
159                  return true;
160                 break;
161             case Qt::Key_Return:
162 
163 
164                  return true;
165                 break;
166 
167             default:
168 
169 
170                 return false;
171 
172                break;
173 
174             }
175 
176         } else  if (event->type() == QEvent::KeyRelease) {
177             QKeyEvent *keyEvent = static_cast<QKeyEvent*>(event);
178 
179 
180             switch (keyEvent->key()) {
181 
182             case Qt::Key_Shift:
183                    emit edit_point_selection(false);
184                   return false;
185                  break;
186 
187             case Qt::Key_Clear:
188                     emit edit_point_deselection(false);
189                     return false;
190                      break;
191 
192 
193             case Qt::Key_Control:
194                     emit edit_point_deselection(false);
195                     return false;
196                      break;
197 
198             case Qt::Key_Delete:
199 
200                     return false;
201                      break;
202             case Qt::Key_Left:
203 
204             case Qt::Key_Right:
205             case Qt::Key_Up:
206             case Qt::Key_Down:
207                 // skip processing
208                 break;
209             case Qt::Key_Enter:
210 
211                  return true;
212                 break;
213             case Qt::Key_Return:
214 
215 
216                  return true;
217                 break;
218 
219             default:
220 
221 
222                 return false;
223 
224                break;
225 
226             }
227 
228         } else {
229             return false;
230         }
231 
232         return false;
233 }
.fi
.SS "void emu_data_editor::selection (boolclear)\fC [slot]\fP"

.PP
Definition at line 340 of file emu_data_editor\&.cpp\&.
.PP
References set_status()\&.
.PP
.nf
341 {
342 
343     emit set_status(clear?2/*deselect action*/:1/*select action*/);
344 
345 }
.fi
.SS "void emu_data_editor::set_status (\fBUint8\fPselect)\fC [signal]\fP"

.PP
Definition at line 296 of file moc_emu_data_editor\&.cpp\&.
.PP
Referenced by selection()\&.
.PP
.nf
297 {
298     void *_a[] = { 0, const_cast<void*>(reinterpret_cast<const void*>(&_t1)) };
299     QMetaObject::activate(this, &staticMetaObject, 7, _a);
300 }
.fi
.SS "void emu_data_editor::set_value (\fBInt16\fPval, \fBUint16\fPwaveform)\fC [signal]\fP"

.PP
Definition at line 270 of file moc_emu_data_editor\&.cpp\&.
.PP
.nf
271 {
272     void *_a[] = { 0, const_cast<void*>(reinterpret_cast<const void*>(&_t1)), const_cast<void*>(reinterpret_cast<const void*>(&_t2)) };
273     QMetaObject::activate(this, &staticMetaObject, 3, _a);
274 }
.fi
.SS "void emu_data_editor::update_curve ()\fC [slot]\fP"

.PP
Definition at line 284 of file emu_data_editor\&.cpp\&.
.PP

AQ0X_TESTER_CONFIG/project/AQ0X/emu_data_editor.cpp
#include "mainwindow.h"
#include "emu_data_editor.h"
#include "ui_emu_data_editor.h"
#include "slider.h"
#include <QFileDialog>
#include <QDesktopServices>
#include <QDateTime>
#include <QMessageBox>
#include <QKeyEvent>


emu_data_editor ::emu_data_editor (QString type,QWidget *parent) :
    QMainWindow(parent),
    ui(new Ui::emu_data_editor )
{

    ui->setupUi(this);
    this->setWindowTitle(type);

    dir_default = ((MainWindow *)parent)->dir_default;

    win_title = type;
    if(type.contains("MAGNETIC FIELD")){
        signal_type = WAVEFORM_PLUS_MINUS_1_65;
        signal_amplitude = 1.65;
    }else{
        signal_type = WAVEFORM_0_TO_3_3;
        signal_amplitude = 3.3;
    }

    signal_freq      = CONFIG_DEFAULT_FREQ;
    signal_phase     = 0;//!defalt values


    ui->centralwidget->setVisible(false);
    ui->dockWidgetCurve_plot->setWidget(ui->plot_curve);
    ui->dockWidgetCurve_plot_fft->setWidget(ui->plot_curve_fft);

    this->addDockWidget(Qt::TopDockWidgetArea,ui->dockWidgetCurve_plot_fft);


    ui->actionWinEditor->setChecked(true);
    ui->actionWinPlot->setChecked(true);



    for (int var = 0; var < EMU_MODE_DATA_LENGTH; ++var) {
        emu_data_xvalue.append(var);
        emu_data_yvalue.append(0);
    }

    plot_curve(emu_data_xvalue,emu_data_yvalue);

    memset(fft_result,0,FFT_SIZE);
    memset(fft_inputs,0,FFT_SIZE);


    transformer.setSize(FFT_SIZE);
    if(!transformer.setWindowFunction("Rectangular"))
        qCritical()<<"unable to set window funtion : emu_data_editor";

    transformer.forwardTransform(fft_inputs,fft_result);


    spectrum_magnitude_max = 0;
    spectrum_magnitude_min = 0;



    for (int var = 0; var < FFT_SIZE; ++var) {

        emu_data_freq_value.append(
                    (var)*signal_freq*
                    ((signal_type == WAVEFORM_PLUS_MINUS_1_65)?1:2));

        qDebug()<<"f("<<var<<") = "<< emu_data_freq_value[var]<<";";
        emu_data_spectrum_magnitude.append((double)fft_result[var]);

        if(fft_result[var] > spectrum_magnitude_max)
            spectrum_magnitude_max = fft_result[var];
        if(fft_result[var] < spectrum_magnitude_min)
            spectrum_magnitude_min = fft_result[var];

    }







    plot_curve_fft(emu_data_freq_value,emu_data_spectrum_magnitude);
    installEventFilter(this);
    gui_init(EMU_MODE_DATA_LENGTH/2);



}


emu_data_editor ::~emu_data_editor ()
{
    delete ui;
}


bool emu_data_editor::eventFilter(QObject *obj, QEvent *event)
{


    if(obj!=this)
        return false;

        if (event->type() == QEvent::KeyPress) {
            QKeyEvent *keyEvent = static_cast<QKeyEvent*>(event);
            switch (keyEvent->key()) {
            case Qt::Key_Shift:
                    emit edit_point_selection(true);
                  return false;
                 break;


            case Qt::Key_Plus:

                    emit adjust_value(10);
                    update_curve();
                    return false;
                 break;

            case Qt::Key_Minus:

                    emit adjust_value(-10);
                    update_curve();
                    return false;
                break;
            case Qt::Key_Left:


                    return false;
                break;

            case Qt::Key_Clear:
                    emit edit_point_deselection(true);
                    return false;
                     break;


            case Qt::Key_Control:
                    emit edit_point_deselection(true);
                    return false;
                 break;


            case Qt::Key_Right:


            case Qt::Key_Enter:

                 return true;
                break;
            case Qt::Key_Return:


                 return true;
                break;

            default:


                return false;

               break;

            }

        } else  if (event->type() == QEvent::KeyRelease) {
            QKeyEvent *keyEvent = static_cast<QKeyEvent*>(event);


            switch (keyEvent->key()) {

            case Qt::Key_Shift:
                   emit edit_point_selection(false);
                  return false;
                 break;

            case Qt::Key_Clear:
                    emit edit_point_deselection(false);
                    return false;
                     break;


            case Qt::Key_Control:
                    emit edit_point_deselection(false);
                    return false;
                     break;

            case Qt::Key_Delete:

                    return false;
                     break;
            case Qt::Key_Left:

            case Qt::Key_Right:
            case Qt::Key_Up:
            case Qt::Key_Down:
                // skip processing
                break;
            case Qt::Key_Enter:

                 return true;
                break;
            case Qt::Key_Return:


                 return true;
                break;

            default:


                return false;

               break;

            }

        } else {
            return false;
        }

        return false;
}



void emu_data_editor::gui_init(Uint16 points)
{
          edit_point_sets.clear();

          for (int var = 0; var < points; ++var) {

           emu_data_edit_point_t edit_point;

           edit_point.val = 0;

           edit_point.gui.slider  = new Slider(var,this);
           edit_point.gui.slider->setMaximum(ADC_RESOLUTION);

           edit_point.gui.vLayout = new QVBoxLayout();
           edit_point.gui.vLayout->addWidget(edit_point.gui.slider);



           edit_point.gui.vLayout->sizeHint();



           edit_point_sets.append(edit_point);

           ui->horizontalLayout_edit_points->addLayout(edit_point.gui.vLayout);
           ui->horizontalLayout_edit_points->sizeHint();




          }


          emit action_set_value(0,SET_TYPE_DEFAULT);

          update_curve();





}





void emu_data_editor ::update_curve(){

     ui->plot_curve->graph(0)->setData(emu_data_xvalue, emu_data_yvalue);

     Int16 head = 0;

     for (int var = 0; var < FFT_SIZE; ++var) {
         fft_inputs[var] = (float)emu_data_yvalue[head];
         head = (head + 1)%EMU_MODE_DATA_LENGTH;
     }

     transformer.forwardTransform(fft_inputs, fft_result);

     emu_data_spectrum_magnitude.clear();


     spectrum_magnitude_max = 0;
     spectrum_magnitude_min = 0;

     for (int var = 0; var < FFT_SIZE; ++var) {

         emu_data_spectrum_magnitude.append((float)fft_result[var]);


         if(fft_result[var] > spectrum_magnitude_max)
             spectrum_magnitude_max = fft_result[var];
         if(fft_result[var] < spectrum_magnitude_min)
             spectrum_magnitude_min = fft_result[var];
     }

#if(0)

     for (int var = 0; var < FFT_SIZE; ++var) {
          qDebug()<<"A("<<var+1<<") = "<< fft_result[var]<<";";
     }

     for (int var = 0; var < FFT_SIZE; ++var) {
          qDebug()<<"B("<<var+1<<") = "<< fft_inputs[var];
     }
#endif



     ui->plot_curve_fft->graph(0)->setData(emu_data_freq_value,emu_data_spectrum_magnitude);
     ui->plot_curve_fft->yAxis->setRange(spectrum_magnitude_min - 2,
                                         spectrum_magnitude_max + 2);



     ui->plot_curve->setColor(QColor("black"));
     ui->plot_curve->replot();
     ui->plot_curve_fft->replot();


}

void emu_data_editor::selection(bool clear)
{

    emit set_status(clear?2/*deselect action*/:1/*select action*/);

}

void emu_data_editor ::plot_curve(QVector<double> xvalue,QVector<double> yvalue)
{

  ui->plot_curve->addGraph();
  ui->plot_curve->setColor(QColor("black"));

  ui->plot_curve->graph(0)->setPen(QPen(Qt::green));
  ui->plot_curve->graph(0)->setData(xvalue, yvalue);

  ui->plot_curve->xAxis->setLabel("t = (n x 100us)");
  ui->plot_curve->yAxis->setLabel("A(t)(V)");

  ui->plot_curve->xAxis->setLabelColor(QColor("black"));
  ui->plot_curve->yAxis->setLabelColor(QColor("black"));

  ui->plot_curve->xAxis->setGridPen(QPen(Qt::gray));
  ui->plot_curve->yAxis->setGridPen(QPen(Qt::gray));
  ui->plot_curve->xAxis2->setGridPen(QPen(Qt::gray));
  ui->plot_curve->yAxis2->setGridPen(QPen(Qt::gray));

  ui->plot_curve->xAxis->setTickPen(QPen(Qt::gray));
  ui->plot_curve->yAxis->setTickPen(QPen(Qt::gray));

  ui->plot_curve->xAxis->setBasePen(QPen(Qt::gray));
  ui->plot_curve->yAxis->setBasePen(QPen(Qt::gray));

  ui->plot_curve->xAxis->setTickLabelColor(QColor("gray"));
  ui->plot_curve->yAxis->setTickLabelColor(QColor("gray"));

  ui->plot_curve->xAxis->setTicks(true);



  ui->plot_curve->xAxis->setSubGrid(true);
  ui->plot_curve->xAxis->setTickStep(1.0);

  ui->plot_curve->xAxis->setSubGridPen(QPen(Qt::gray));
  ui->plot_curve->yAxis->setSubGridPen(QPen(Qt::gray));

  ui->plot_curve->xAxis2->setVisible(true);
  ui->plot_curve->xAxis2->setTickLabels(false);
  ui->plot_curve->yAxis2->setVisible(true);
  ui->plot_curve->yAxis2->setTickLabels(false);
  ui->plot_curve->xAxis2->setBasePen(QPen(Qt::gray));
  ui->plot_curve->yAxis2->setBasePen(QPen(Qt::gray));

   ui->plot_curve->xAxis->setGrid(true);
   ui->plot_curve->yAxis->setGrid(true);


  ui->plot_curve->xAxis->setRange(0, EMU_MODE_DATA_LENGTH);
  if(this->signal_type == WAVEFORM_PLUS_MINUS_1_65)
  ui->plot_curve->yAxis->setRange(-1.65, 1.65);
  else
  ui->plot_curve->yAxis->setRange(0, 3.34);
}



void emu_data_editor ::plot_curve_fft(QVector<double> xvalue,QVector<double> yvalue)
{


  ui->plot_curve_fft->addGraph();
  ui->plot_curve_fft->setColor(QColor("black"));

  ui->plot_curve_fft->graph(0)->setPen(QPen(Qt::green));
  ui->plot_curve_fft->graph(0)->setData(xvalue, yvalue);

  ui->plot_curve_fft->xAxis->setLabel("f ");
  ui->plot_curve_fft->yAxis->setLabel("A(f)(V)");

  ui->plot_curve_fft->xAxis->setLabelColor(QColor("black"));
  ui->plot_curve_fft->yAxis->setLabelColor(QColor("black"));

  ui->plot_curve_fft->xAxis->setGridPen(QPen(Qt::gray));
  ui->plot_curve_fft->yAxis->setGridPen(QPen(Qt::gray));
  ui->plot_curve_fft->xAxis2->setGridPen(QPen(Qt::gray));
  ui->plot_curve_fft->yAxis2->setGridPen(QPen(Qt::gray));

  ui->plot_curve_fft->xAxis->setTickPen(QPen(Qt::gray));
  ui->plot_curve_fft->yAxis->setTickPen(QPen(Qt::gray));

  ui->plot_curve_fft->xAxis->setBasePen(QPen(Qt::gray));
  ui->plot_curve_fft->yAxis->setBasePen(QPen(Qt::gray));

  ui->plot_curve_fft->xAxis->setTickLabelColor(QColor("gray"));
  ui->plot_curve_fft->yAxis->setTickLabelColor(QColor("gray"));

  ui->plot_curve_fft->xAxis->setTicks(true);



  ui->plot_curve_fft->xAxis->setSubGrid(true);
  ui->plot_curve_fft->xAxis->setTickStep(1
                                         );

  ui->plot_curve_fft->xAxis->setSubGridPen(QPen(Qt::gray));
  ui->plot_curve_fft->yAxis->setSubGridPen(QPen(Qt::gray));

  ui->plot_curve_fft->xAxis2->setVisible(true);
  ui->plot_curve_fft->xAxis2->setTickLabels(false);
  ui->plot_curve_fft->yAxis2->setVisible(true);
  ui->plot_curve_fft->yAxis2->setTickLabels(false);
  ui->plot_curve_fft->xAxis2->setBasePen(QPen(Qt::gray));
  ui->plot_curve_fft->yAxis2->setBasePen(QPen(Qt::gray));

   ui->plot_curve_fft->xAxis->setGrid(true);
   ui->plot_curve_fft->yAxis->setGrid(true);



  ui->plot_curve_fft->xAxis->setRange(0,
                                       2000);
  ui->plot_curve_fft->yAxis->setRange(spectrum_magnitude_min - 2,
                                      spectrum_magnitude_max + 2);
}



void emu_data_editor ::saveEventRecored(QString FileName){

    QString EventRecordFileName;

            QDir default_dir;

            if(!default_dir.exists(default_dir.toNativeSeparators(dir_default)))
                default_dir.mkdir(default_dir.toNativeSeparators(dir_default));

            EventRecordFileName = default_dir.toNativeSeparators(dir_default)+ EVENT_RECORD_FILE;


       QFile outputFile(EventRecordFileName);

       if(outputFile.exists())
       outputFile.open(QIODevice::WriteOnly | QIODevice::Append);
       else
       outputFile.open(QIODevice::WriteOnly);

       if(!outputFile.isOpen()){
           ((MainWindow *)parent())->msg(1,"Error, unable to open " + EventRecordFileName + " ");

            return;
       }

       QTextStream outStream(&outputFile);

       outStream << FileName<<"\n";
       outputFile.close();
}





void emu_data_editor::action_set_value_triggered(double val, Uint8 type)
{

    if(type == SET_TYPE_SIGNAL_FREQ)
        signal_freq = val;
    else if(type == SET_TYPE_PHASE_SHIFT)
        signal_phase = val/M_PI;
    else if(type == SET_TYPE_SIGNAL_AMP)
        signal_amplitude = val;

    emit action_set_value(val,type);
    update_curve();
}



void emu_data_editor::on_actionWinEditor_toggled(bool arg1)
{
     ui->actionWinEditor->setChecked(arg1);

     if(arg1)
         ui->dockWidgetCurve_plot->show();
     else
         ui->dockWidgetCurve_plot->close();

}

void emu_data_editor::on_actionWinPlot_toggled(bool arg1)
{
     ui->actionWinPlot->setChecked(arg1);

     if(arg1)
         ui->dockWidgetCurve_edit->show();
     else
         ui->dockWidgetCurve_edit->close();
}

void emu_data_editor::on_actionOpen_triggered()
{
    QDir   default_dir;
    QString SelectedFilesName;


    if(default_dir.exists(default_dir.toNativeSeparators(dir_default)))
        default_dir.mkdir(default_dir.toNativeSeparators(dir_default));




    SelectedFilesName = QFileDialog::getOpenFileName(
        this,
        tr("Open File"),
        "",
        tr("Comma Separated Values (*.csv)"));


     loadFile(SelectedFilesName);

}

void emu_data_editor::loadFile(QString file_name){

    bool   sucess = true;

    if (file_name.isEmpty()) return;

        QFile file(file_name);

        if (!file.open(QIODevice::ReadOnly)) {
            QMessageBox::information(this, tr("Unable to open file"), file.errorString());
            return;
        }


QTextStream stream(&file);
Uint16 arg_cnt;
unsigned int time;


       QString line = stream.readLine();


       for(unsigned int ii=0;ii<(EMU_MODE_DATA_LENGTH/2);ii++){

             line = stream.readLine();

             arg_cnt = sscanf(line.toLatin1(),"%u%*1c%u",&time,&(edit_point_sets[ii].val)
                              );


             if(arg_cnt!=2 ||time!= ii){

                 QMessageBox::StandardButton reply;
                 reply = QMessageBox::question(this, "Error Report",
                                               "This file is corrupted, do you want to open anyway",
                                               QMessageBox::Yes|QMessageBox::No);
                 if (reply == QMessageBox::No)
                  sucess = false;
                  break;
             }else {

                  edit_point_sets[ii].gui.slider->setValue(edit_point_sets[ii].val);
             }
             }

       stream.flush();
       file.close();





       emit update_value();
       update_curve();


}


void emu_data_editor::on_actionCSV_triggered()
{

    QDir default_dir;



    if(default_dir.exists(default_dir.toNativeSeparators(dir_default)))
        default_dir.mkdir(default_dir.toNativeSeparators(dir_default));

    QString DefaultFileName =  default_dir.toNativeSeparators(dir_default)+(win_title
                                                                            +"_File_Saved_At_" +
                               QDateTime::currentDateTime().toString(Qt::SystemLocaleLongDate)
                               );


    DefaultFileName.replace(':',"-").replace(' ',"_");


    QString SelectedFilesName = QFileDialog::getSaveFileName(
        this,
        tr("Save As"),
        DefaultFileName,
        tr("Comma Separated Values (*.csv)"));


    if(SelectedFilesName == NULL)
        return;

     fileName = SelectedFilesName;
    if (fileName.isEmpty()) return;
    else {
        QFile file(fileName);
        if (!file.open(QIODevice::WriteOnly)) {
            QMessageBox::information(this, tr("Unable to open file"), file.errorString());
            return;
        }

        QTextStream stream(&file);

        stream << "TIME, DATA\n";

        for(int ii=0;ii < (EMU_MODE_DATA_LENGTH/2)/*data edit points*/;ii++)
           stream << QString::number(ii) +","+QString::number(edit_point_sets[ii].gui.slider->value())+"\n";



        file.close();



        try {
            QDesktopServices::openUrl(QUrl("file:"+ fileName));

        } catch (...) {

        }
    }


}



void emu_data_editor::on_btn_save_clicked()
{

    Uint16 data[EMU_MODE_DATA_LENGTH];
    double adjust = windowTitle().contains("MAGNETIC FIELD")?1.65:0;


    for (int var = 0; var < emu_data_yvalue.length(); ++var) {
         data[var]=((emu_data_yvalue[var]+adjust)/ADC_VAL_MAPPING_FACTOR);
    }

    ((MainWindow *)this->parent())->set_emu_data(this,data,sizeof(data));

}

void emu_data_editor::on_actionMakeImg_triggered()
{

    QDir default_dir;

    if(!default_dir.exists(default_dir.toNativeSeparators(dir_default)))
        default_dir.mkdir(default_dir.toNativeSeparators(dir_default));

    QString DefaultFileName = this->windowTitle();



    DefaultFileName.replace(':',"-").replace(' ',"_").replace('.',"_");

    DefaultFileName = default_dir.toNativeSeparators(dir_default)+ DefaultFileName;

    QString SelectedFilesName = QFileDialog::getSaveFileName(this, tr("Save File"),
                                                        DefaultFileName,
                                                         tr("Images (*.png *.jpg *.bmp)"));

    if(SelectedFilesName == NULL)
        return;

    ui->plot_curve->graph(0)->setPen(QPen(Qt::green));
    /*
    ui->plot_curve->graph(1)->setPen(QPen(Qt::red));
    ui->plot_curve->graph(1)->setPen(QPen(Qt::DotLine));
    ui->plot_curve->graph(1)->setBrush(QBrush(Qt::gray));
    */

    ui->plot_curve->setColor(QColor("white"));
    ui->plot_curve->replot();




    if(SelectedFilesName.indexOf('.')==-1)
        SelectedFilesName.append(".jpg");

    if((SelectedFilesName.indexOf('.')==SelectedFilesName.length()-4)){
    if(SelectedFilesName.endsWith(".jpg")){
       if(
        !ui->plot_curve->saveJpg(SelectedFilesName
                                ,800,600))
            ((MainWindow *)parent())->msg(MSG_ERROR,"Error: Couldn't save data");
    }else

    if(SelectedFilesName.endsWith(".png")){
           if(
            !ui->plot_curve->savePng(SelectedFilesName
                                    ,800,600))
                ((MainWindow *)parent())->msg(MSG_ERROR,"Error: Couldn't save data");

    }else

    if(SelectedFilesName.endsWith(".bmp")){
                  if(
                   !ui->plot_curve->saveBmp(SelectedFilesName
                                           ,800,600))
                       ((MainWindow *)parent())->msg(MSG_ERROR,"Error: Couldn't save data");

     }

    }
     ui->plot_curve->graph(0)->setPen(QPen(Qt::green));

     /*
     ui->plot_curve->graph(1)->setPen(QPen(Qt::yellow));
     ui->plot_curve->graph(1)->setBrush(QBrush(Qt::yellow));
     */
     ui->plot_curve->setColor(QColor("black"));
     ui->plot_curve->replot();

       if(!QDesktopServices::openUrl(QUrl(SelectedFilesName)))
           ((MainWindow *)parent())->msg(MSG_ERROR,"Couldn't open the file"+SelectedFilesName);

}
